blog/_posts/2023-01-16-wsl2spackapptainer.md

```diff
@@ -0,0 +1,289 @@
+
+
+class PyEspressoWalberla(CMakePackage):
+    """ESPResSo is a highly versatile software package for performing and
+    analyzing scientific Molecular Dynamics many-particle simulations of
+    coarse-grained atomistic or bead-spring models as they are used in
+    soft matter research in physics, chemistry and molecular biology. It
+    can be used to simulate systems such as polymers, liquid crystals,
+    colloids, polyelectrolytes, ferrofluids and biological systems, for
+    example DNA and lipid membranes. It also has a DPD and lattice
+    Boltzmann solver for hydrodynamic interactions, and allows several
+    particle couplings to the LB fluid.
+    """
+
+    homepage = "https://espressomd.org/"
+    git = "https://github.com/RudolfWeeber/espresso"
+
+    version("walberla", branch="walberla")
+
+    depends_on("cmake@3.0:", type="build")
+    depends_on("mpi")
+    depends_on("boost+serialization+filesystem+system+python+mpi")
+
+    # TODO: replace this with an explicit list of components of Boost,
+    # for instance depends_on('boost +filesystem')
+    # See https://github.com/spack/spack/pull/22303 for reference
+    depends_on(Boost.with_default_variants)
+    extends("python")
+    depends_on("py-cython@0.23:", type="build")
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("fftw")
+    depends_on("hdf5+hl+mpi")
+```
+
+That includes everything but turning on the waLBerla support, so we add in:
+
+```python
+    variant("walberla", default=True, description="Include waLBerla support")
+    def cmake_args(self):
+        spec = self.spec
+        args = [
+            self.define_from_variant("ESPRESSO_BUILD_WITH_WALBERLA", "walberla")
+        ]
+        return args
+```
+
+Have we arrived at working build?  `spack install py-espresso-walberla` rather does make it look like we have.
+
+Now was that really the right way to add this new version of py-espresso?  I'm not going to lose sleep over it right now.
+
+## Putting this in a container
+
+What if I wanted to export this as a container to run on another system?  This is where I think we can plug Apptainer and Spack together.
+
+### A basic Spack container recipe
+
+Spack has support for making Apptainer compatible recipes for exporting Spack environments into containers, but it's not quite up to the challenge of working with this modified Spack (that exists on this machine alone).
+
+So let's do the nearly-but-not quite, then tweak it.  If we write a spack.yaml file:
+```yaml
+spack:
+  specs:
+  - py-espresso-walberla
+
+  container:
+    format: singularity
+    images:
+      os: "ubuntu:22.04"
+      spack: develop
+```
+We can then ask spack to generate a recipe:
+```bash
+spack containerize > py-espresso-walberla.def
+```
+
+That tries to use Spack from within their Docker container, not our version which contains the extra packages we've added, so let's just tweak it to delete the version in the container, and copy ours in instead.
```

Comment:
Sorry nit picking now, but assume it uses a base Docker image that contains the standard Spack you're referring to?

Reply:
Yes it does.  You want me to be more explicit explaining that's how it works?

Reply:
Yeah I think explicitly saying that might be nice for people uncertain of how Spack is creating the container recipe.

Reply:
Extra explanation added in e1480f9

I'll release this now.